vpc/tool/container2/setup_container_linux.go

```diff
@@ -254,52 +249,169 @@ func configureLink(ctx context.Context, nsHandle *netlink.Handle, link netlink.L
+	ctx, cancel := context.WithTimeout(ctx, networkSetupWait)
+	defer cancel()
+
+	runtime.LockOSThread()
+	defer runtime.UnlockOSThread()
+
+	// Amazon only gives out /128s
+	new6IP := net.IPNet{IP: net.ParseIP(allocation.IPV6Address.Address.Address), Mask: net.CIDRMask(128, 128)}
+	new6Addr := netlink.Addr{
+		// TODO (Sargun): Check IP Mask setting.
+		IPNet: &new6IP,
+		Flags: unix.IFA_F_PERMANENT | unix.IFA_F_NODAD,
+		Scope: unix.RT_SCOPE_UNIVERSE,
+	}
+	err := nsHandle.AddrAdd(link, &new6Addr)
 	if err != nil {
-		return false, err
+		logger.G(ctx).WithError(err).Error("Unable to add IPv6 addr to link")
+		return errors.Wrap(err, "Unable to add IPv6 addr to link")
 	}
-	for _, addr := range addrs {
-		if addr.Scope == int(netlink.SCOPE_UNIVERSE) && addr.Flags&unix.IFA_F_TENTATIVE == 0 {
-			goto addr_found
+
+	var linkLocalAddr netlink.Addr
+	for {
+		// We need *a* permanent address
+		addrs, err := nsHandle.AddrList(link, netlink.FAMILY_V6)
+		if err != nil {
+			return fmt.Errorf("Cannot get addrs: %w", err)
+		}
+
+		for _, addr := range addrs {
+			if (addr.Scope == unix.RT_SCOPE_LINK) && (addr.Flags&(unix.IFA_F_TENTATIVE|unix.IFA_F_PERMANENT) == unix.IFA_F_PERMANENT) {
+				linkLocalAddr = addr
+				goto out
+			}
 		}
+		select {
+		case <-ctx.Done():
+			return fmt.Errorf("Did not find permanent link-local address in set: %v: %w", addrs, ctx.Err())
+		case <-time.After(100 * time.Millisecond):
+		}
+	}
+
+out:
+	origns, err := netns.Get()
+	defer origns.Close()
+	if err != nil {
+		return fmt.Errorf("Could not get current net ns: %w", err)
 	}
-	return false, nil
-addr_found:
-	_, err = nsHandle.RouteGet(net.ParseIP("2001::1"))
-	if err == unix.ENETUNREACH {
-		return false, nil
-	} else if err != nil {
-		return false, err
+
+	err = netns.Set(netns.NsHandle(netnsfd))
+	if err != nil {
+		return fmt.Errorf("Could not switch net ns: %w", err)
 	}
-	return true, nil
-}
 
-func waitForAddressUp(ctx context.Context, nsHandle *netlink.Handle, link netlink.Link) error {
-	ctx, cancel := context.WithTimeout(ctx, networkSetupWait)
-	defer cancel()
-	pollTimer := time.NewTicker(100 * time.Millisecond)
-	defer pollTimer.Stop()
+	networkInterface, err := net.InterfaceByName("eth0")
+	if err != nil {
+		return fmt.Errorf("Could not get network interface eth0: %w", err)
+	}
+
+	conn, _, err := ndp.Dial(networkInterface, ndp.Addr(linkLocalAddr.IP.String()))
+	if err != nil {
+		return fmt.Errorf("Unable to dial NDP conn: %w", err)
+	}
+	defer conn.Close()
 
+	var filter ipv6.ICMPFilter
+	filter.SetAll(true)
+	filter.Accept(ipv6.ICMPTypeRouterAdvertisement)
+	err = conn.SetICMPFilter(&filter)
+	if err != nil {
+		return fmt.Errorf("Unable to setup ICMP filter %v: %w", filter, err)
+	}
+
+	m := &ndp.RouterSolicitation{
+		Options: []ndp.Option{
+			&ndp.LinkLayerAddress{
+				Direction: ndp.Source,
+				Addr:      networkInterface.HardwareAddr,
+			},
+		},
+	}
+
+	/* Poll for the router */
```

Comment:
What do you think about a function for this called waitForIPV6NDPAdvertisement ?